BlueprintUICommonControls/Sources/Internal/KeyboardObserver.swift

```diff
@@ -0,0 +1,166 @@
+        options : UIView.AnimationOptions
+    )
+}
+
+/**
+ Implementation borrowed from Listable:
+ https://github.com/kyleve/Listable/blob/master/Listable/Sources/Internal/KeyboardObserver.swift
+ 
+ iOS Docs for keyboard management:
+ https://developer.apple.com/library/archive/documentation/StringsTextFonts/Conceptual/TextAndWebiPhoneOS/KeyboardManagement/KeyboardManagement.html
+ */
+final class KeyboardObserver {
+    
+    private let center : NotificationCenter
+    
+    weak var delegate : KeyboardObserverDelegate?
+    
+    //
+    // MARK: Initialization
+    //
+    
+    init(center : NotificationCenter = .default) {
+        
+        self.center = center
+
+        self.center.addObserver(self, selector: #selector(keyboardFrameChanged(_:)), name: UIWindow.keyboardWillChangeFrameNotification, object: nil)
+        self.center.addObserver(self, selector: #selector(keyboardFrameChanged(_:)), name: UIWindow.keyboardDidChangeFrameNotification, object: nil)
+    }
+    
+    private var latestNotification : NotificationInfo?
+    
+    //
+    // MARK: Handling Changes
+    //
+    
+    enum KeyboardFrame : Equatable {
+        
+        case notVisible
+        case visible(frame: CGRect)
+    }
+    
+    func currentFrame(in view : UIView) -> KeyboardFrame? {
+        
+        guard view.window != nil else {
+            return nil
+        }
+        
+        guard let notification = self.latestNotification else {
+            return nil
+        }
+        
+        let frame = view.convert(notification.endingFrame, from: nil)
+        
+        if frame.intersects(view.bounds) {
+            return .visible(frame: frame)
+        } else {
+            return .notVisible
+        }
+    }
+    
+    //
+    // MARK: Receiving Updates
+    //
+    
+    private func receivedUpdatedKeyboardInfo(_ new : NotificationInfo) {
+        
+        let old = self.latestNotification
+        
+        self.latestNotification = new
+        
+        /// Only communicate a frame change to the delegate if the frame actually changed.
+        
+        if let old = old {
+            guard old.endingFrame != new.endingFrame else {
+                return
+            }
+        }
+        
+        /**
+         Create an animation curve with the correct curve for showing or hiding the keyboard.
+         
+         This is unfortunately a private UIView curve. However, we can map it to the animation options' curve
+         like so: https://stackoverflow.com/questions/26939105/keyboard-animation-curve-as-int
+         */
+        let animationOptions = UIView.AnimationOptions(rawValue: new.animationCurve << 16)
+        
+        self.delegate?.keyboardFrameWillChange(
+            for: self,
+            animationDuration: new.animationDuration,
+            options: animationOptions
+        )
+    }
+    
+    //
+    // MARK: Notification Listeners
+    //
+    
+    @objc func keyboardFrameChanged(_ notification : Notification) {
```

Comment:
Any reason not to make this private?
```suggestion
    @objc private func keyboardFrameChanged(_ notification : Notification) {
```

Reply:
TIL you can directly suggest changes in Github

Reply:
Landing this locally